internal/controller/supervisorconfig/upstreamwatcher/upstreamwatcher.go

```diff
@@ -0,0 +1,320 @@
+	reasonNotFound             = "SecretNotFound"
+	reasonWrongType            = "SecretWrongType"
+	reasonMissingKeys          = "SecretMissingKeys"
+	reasonSuccess              = "Success"
+	reasonUnreachable          = "Unreachable"
+	reasonInvalidResponse      = "InvalidResponse"
+
+	// Errors that are generated by our reconcile process.
+	errFailureStatus = constable.Error("UpstreamOIDCProvider has a failing condition")
+)
+
+// IDPCache is a thread safe cache that holds a list of validated upstream OIDC IDP configurations.
+type IDPCache interface {
+	SetIDPList([]provider.UpstreamOIDCIdentityProvider)
+}
+
+type controller struct {
+	cache          IDPCache
+	log            logr.Logger
+	client         pinnipedclientset.Interface
+	providers      idpinformers.UpstreamOIDCProviderInformer
+	secrets        corev1informers.SecretInformer
+	validatorCache *cache.Expiring
+}
+
+// New instantiates a new controllerlib.Controller which will populate the provided IDPCache.
+func New(
+	idpCache IDPCache,
+	client pinnipedclientset.Interface,
+	providers idpinformers.UpstreamOIDCProviderInformer,
+	secrets corev1informers.SecretInformer,
+	log logr.Logger,
+) controllerlib.Controller {
+	c := controller{
+		cache:          idpCache,
+		log:            log.WithName(controllerName),
+		client:         client,
+		providers:      providers,
+		secrets:        secrets,
+		validatorCache: cache.NewExpiring(),
+	}
+	filter := pinnipedcontroller.MatchAnythingFilter(pinnipedcontroller.SingletonQueue())
+	return controllerlib.New(
+		controllerlib.Config{Name: controllerName, Syncer: &c},
+		controllerlib.WithInformer(providers, filter, controllerlib.InformerOption{}),
+		controllerlib.WithInformer(secrets, filter, controllerlib.InformerOption{}),
+	)
+}
+
+// Sync implements controllerlib.Syncer.
+func (c *controller) Sync(ctx controllerlib.Context) error {
+	actualUpstreams, err := c.providers.Lister().List(labels.Everything())
+	if err != nil {
+		return fmt.Errorf("failed to list UpstreamOIDCProviders: %w", err)
+	}
+
+	requeue := false
+	validatedUpstreams := make([]provider.UpstreamOIDCIdentityProvider, 0, len(actualUpstreams))
+	for _, upstream := range actualUpstreams {
+		valid := c.validateUpstream(ctx, upstream)
+		if valid == nil {
+			requeue = true
+		} else {
+			validatedUpstreams = append(validatedUpstreams, *valid)
+		}
+	}
+	c.cache.SetIDPList(validatedUpstreams)
+	if requeue {
+		return controllerlib.ErrSyntheticRequeue
+	}
+	return nil
+}
+
+// validateUpstream validates the provided v1alpha1.UpstreamOIDCProvider and returns the validated configuration as a
+// provider.UpstreamOIDCIdentityProvider. As a side effect, it also updates the status of the v1alpha1.UpstreamOIDCProvider.
+func (c *controller) validateUpstream(ctx controllerlib.Context, upstream *v1alpha1.UpstreamOIDCProvider) *provider.UpstreamOIDCIdentityProvider {
+	scopes := append([]string{"openid"}, upstream.Spec.AuthorizationConfig.AdditionalScopes...)
```

Comment:
uh what happens if they ask for `openid` in additional scopes?

Reply:
Oh, yeah maybe we should de-dupe as well. I think most upstream providers probably behave reasonably if you did send a duplicate scope, but we can eliminate that case pretty easily so we might as well.